File: application/src/main/java/com/example/movie/MovieService.java
package com.example.movie;

import com.example.movie.request.MovieSearchServiceRequest;
import com.example.movie.response.MovieResponse;
import com.example.movie.response.MoviesServiceResponse;
import com.example.repository.movie.MovieRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Slf4j
@Service
@RequiredArgsConstructor
public class MovieService {

    private final MovieRepository movieRepository;

    @Transactional(readOnly = true)
    @Cacheable(value = "movies", key = "#request.title + ':' + #request.genre")
    public List<MovieResponse> getMovies(MovieSearchServiceRequest request) {
        return MoviesServiceResponse.of(movieRepository.findAllByTitleAndGenre(request.getTitle(), request.getGenre()));
    }
}

File: module-application/src/main/kotlin/com/example/redis/movie/in/movie/MovieService.kt
package com.example.redis.movie.`in`.movie

import com.example.redis.movie.Movie
import com.example.redis.movie.Reservation
import com.example.redis.movie.out.movie.MoviePort
import com.example.redis.movie.out.movie.TheaterPort
import org.springframework.cache.annotation.Cacheable
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class MovieService(
    private val movieAdapter: MoviePort,
    private val theaterAdapter: TheaterPort,
): MovieUseCase {

    @Cacheable(value = ["movieSearchCache"], keyGenerator = "movieSearchKeyGenerator")
    @Transactional(readOnly = true)
    override fun gets(title: String?, genre: String?): MutableList<Movie> {
        val movies = movieAdapter.findByOrderByReleaseDateDesc(title, genre)
        return movies
    }

    @Transactional(readOnly = false)
    override fun reserve(movieId: Long, reservation: Reservation): String {
        val id = movieAdapter.reserve(reservation)
        return id
    }
}